scripts/canaryRelease.mts

```diff
@@ -0,0 +1,242 @@
+}
+
+async function listPublicChangedPackages(baseline: string) {
+  const { stdout: packagesJson } =
+    await $`pnpm list --recursive --filter ...[${baseline}] --depth -1 --only-projects --json`;
+  const packages = JSON.parse(packagesJson) as PackageInfo[];
+  return packages.filter((pkg) => !pkg.private);
+}
+
+async function getChangedPackages(
+  baseline: string | undefined,
+  skipLastCommitComparison: boolean,
+): Promise<PackageInfo[]> {
+  if (!skipLastCommitComparison) {
+    const publicPackagesUpdatedInLastCommit = await listPublicChangedPackages('HEAD~1');
+    if (publicPackagesUpdatedInLastCommit.length === 0) {
+      console.log('No public packages changed in the last commit.');
+      return [];
+    }
+  }
+
+  if (!baseline) {
+    const { stdout: latestTag } = await $`git describe --abbrev=0`;
+    baseline = latestTag;
+  }
+
+  console.log(`Looking for changed public packages since ${chalk.yellow(baseline)}...`);
+
+  const changedPackages = await listPublicChangedPackages(baseline);
+  if (changedPackages.length === 0) {
+    console.log('Nothing found.');
+  }
+
+  return changedPackages;
+}
+
+async function confirmPublishing(changedPackages: PackageInfo[], yes: boolean) {
+  if (!yes) {
+    const rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
+
+    console.log('\nFound changes in the following packages:');
+    for (const pkg of changedPackages) {
+      console.log(` - ${pkg.name}`);
+    }
+
+    console.log('\nThis will publish the above packages to the npm registry.');
+    const answer = await rl.question('Do you want to proceed? (y/n) ');
+
+    rl.close();
+
+    if (answer.toLowerCase() !== 'y') {
+      console.log('Aborted.');
+      process.exit(0);
+    }
+  }
+}
+
+async function setAccessToken(npmAccessToken: string | undefined) {
+  if (!npmAccessToken && !process.env.NPM_TOKEN) {
+    console.error(
+      '❌ NPM access token is required. Either pass it as an argument or set it as an NPM_TOKEN environment variable.',
+    );
+    process.exit(1);
+  }
+
+  const npmrcPath = resolve(workspaceRoot, '.npmrc');
+
+  await appendFile(
+    npmrcPath,
+    `//registry.npmjs.org/:_authToken=${npmAccessToken ?? process.env.NPM_TOKEN}\n`,
+  );
+}
+
+async function setVersion(packages: PackageInfo[]) {
+  const { stdout: currentRevisionSha } = await $`git rev-parse --short HEAD`;
+  const timestamp = formatDate(new Date());
+  let hasError = false;
+
+  const tasks = packages.map(async (pkg) => {
+    const packageJsonPath = resolve(pkg.path, './package.json');
+    try {
+      const packageJson = JSON.parse(await readFile(packageJsonPath, { encoding: 'utf8' }));
+      const version = packageJson.version;
+      const dashIndex = version.indexOf('-');
+      let newVersion = version;
+      if (dashIndex !== -1) {
+        newVersion = version.slice(0, dashIndex);
+      }
+
+      newVersion = `${newVersion}-dev.${timestamp}-${currentRevisionSha}`;
```

Comment:
In the format, is it possible to drop the time from the date? https://www.npmjs.com/package/typescript?activeTab=versions seems clearer and good enough since we have the commit.

Reply:
Actually, maybe even?

6.0.0-dev-2024-04-24-9968b4889d

would be clearer.

Reply:
This was done to ensure the version suffix is always increasing. In the TypeScript case, it's OK, as they publish one version a day. If we're targeting to publish each commit, having time is required to ensure increasing versions.